### PR TITLE
Finetune execution for 0.4.0 release

### PR DESCRIPTION
**Pull request by gabor-boros**
**Reason for the change**

The dictionary traversal is fishy and preconditions reporting errors instead of warning since they are part of the execution chain.

**Description**

* New exception type ``PreconditionFailedError`` indicating that the precondition failed and no need to raise an error
* "No changes made by" messages now info logs instead of warnings
* Commit changes only if the Law has passing rules
* If ``PreconditionFailedError`` raised, do not log error messages, log a warning instead
* ``LineExists`` will not raise an exception if multiple targets found, instead it will select the last match as target
* Fixed nested rule indentation PR description markup

Also, fixing a typo in the changelog.

**Code examples**

N/A

**Checklist**

- [x] Unit tests created/updated
- [x] Documentation extended/updated

**References**

N/A

### CODE REVIEW

File: hammurabi/rules/text.py [hunk 1]
```diff
@@ -155,7 +152,9 @@ def task(self) -> Path:
 
         if not file_was_empty and no_criteria_match:
             target_match = self.__get_target_match(lines)
-            target_match_index = lines.index(target_match)
+
+            # Get the index of the element from the right
```
Comment (balazs-tothfalussy) [inline hammurabi/rules/text.py:156]: why is this needed? this changes semantics a lot, how did it work before?

**Review by balazs-tothfalussy — APPROVED** [review]

File: hammurabi/rules/text.py (hunk 1 above)
Comment (gabor-boros) [reply hammurabi/rules/text.py:156]: Before this change, the LineExists raised an exception if multiple matches found. This would not be a bad idea, but I did not count with the case: What if we want to append something at the end of the file?

What happens now (and before) is that we read the file, and split it by lines, which is good for us, because we can iterate every line, check for matches and so on. The only (and huge) issue was that we were not able to add something to the beginning or end of the file.

This change will ensure that we will insert after the last match if we found multiple matches for the target. This way we can insert anything to the end of the file (which, again, is a list of splitted lines) 😊

Comment (balazs-tothfalussy) [reply hammurabi/rules/text.py:156]: hm, right
